WIP: vmware dynamic inventory naming collision

agowa opened the pull request:
##### SUMMARY
Possible name collision within the vmware dynamic inventory plugin.
This change vmname to a urlencoded vmpath (to make it unique).
Fixes #55135 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/plugins/inventory/vmware_vm_inventory.py

##### ADDITIONAL INFORMATION
Currently the inventory name is the vmname concatenated with it's uuid. But neither the name nor the uuid is guaranteed to be unique (both can be set arbitrarily). The only thing that is unique, is the full folder path including the vm name. Within a single folder, it is not allowed (and not possible) to have two vms with identical name.
Also special chars are not escaped properly. VMware allows any character within vm names, folder names, ... This causes even trouble within VMwares own PowerCLI api (the powershell one)

##### DRAFT Status
Currently this pr is a draft, as it slows down the inventory plugin enormously.

dgmorales committed:
VMware: more tests for vmware_vm_inventory

Improved integration test for testing for folder support features.

agowa committed:
Fix #55135 vmware dynamic inventory naming collision

Akasurde commented on lib/ansible/plugins/inventory/vmware_vm_inventory.py:
```diff
@@ -450,6 +456,23 @@ def _populate_from_source(self, source_data, using_current_cache):
                     cacheable_results[vm_guest_id]['hosts'].append(current_host)
                     self.inventory.add_child(vm_guest_id, current_host)
 
+                    # Based on folder of virtual machine
```
Please see this - https://github.com/ansible/ansible/pull/52782

agowa replied:
Oh, sorry, that should not have been there. Didn't notice, that I accidentally committed that change too...

agowa committed:
Fix TRANSFORM_INVALID_GROUP_CHARS
FIX: [DEPRECATION WARNING]: The TRANSFORM_INVALID_GROUP_CHARS settings is set to allow bad characters in group names by default, this will
change, but still be user configurable on deprecation. This feature will be removed in version 2.10. Deprecation warnings can be disabled
by setting deprecation_warnings=False in ansible.cfg.

agowa committed:
Make python2 compatible

agowa committed:
Add documentation and mirror vmwares path schema from PowerCLI

agowa committed:
Move documentation comment into docstring

mattclay commented:
bot_status

agowa338 commented:
bot_status

jillr reviewed (changes requested):
Thanks for this patch @agowa338.  This creates a big change in the inventory output which could cause problems for some folks.  We'd rather see this as an option, defaulting to False to preserve the existing behaviour.  Something like generate_unique_names=False|True, and use new the code on get_option()

agowa committed:
Add naming_schema config options

agowa338 commented:
rerun ci

agowa committed:
Fix quoting

agowa338 commented:
bot_status

agowa338 commented:
ready_for_review

jillr reviewed (commented):
Thanks @agowa338, this looks good.  Would you be able to add integration tests for the new naming formats?

agowa338 commented:
I actually don't know. I haven't done that much with integration tests until now, but I'll have a look.

jillr commented:
Thanks! It would generally mean adding a new test in `test/integration/targets/inventory_vmware_vm_inventory/` for the plugin config to enable this option and updating the `runme.sh`.  Some of the other inventory plugin tests have examples of testing multiple different configs so there should be some good examples to compare to.

agowa committed:
Fix integration test silently failing + add debug output

agowa committed:
Integrationtest: Add folders with special characters

agowa committed:
Add integration test for naming schema

agowa committed:
Fix quotation

jillr commented on test/integration/targets/inventory_vmware_vm_inventory/runme.sh:
```diff
@@ -12,30 +13,73 @@ export VMWARE_SERVER="${VCENTER_HOST}"
+# This is required, as otherwise we overwite exit codes of commands we call and the integration
+# test fails prematurely, but reports successful execution
+failure_cleanup() {
+    echo "Previous command faild."
```
```suggestion
    echo "Previous command failed."
```

jillr reviewed (commented):
Thanks very much for this, really good start.  Tests are not passing for me though.  With govcsim it fails out on the yaml test with a parse error.  On our baremetal lab (which is still in development but tests have been ported to support it) - `VCENTER_HOST` is not set in the env.  There's a `prepare_vmware_tests` role you can use that will give you reliable vars for use in vmware testing.
https://github.com/ansible/ansible/tree/devel/test/integration/targets/prepare_vmware_tests/vars

I'm updating the vmware guide in #60163 to cover use of this role if you would like to add a review.

Commit:
Update test/integration/targets/inventory_vmware_vm_inventory/runme.sh

Co-Authored-By: Jill R <4121322+jillr@users.noreply.github.com>

agowa committed:
don't overwrite urllib, as this breaks the vapi module (they erroneously use urlencode to encode the hostname), therefore tags were broken.
